# Magma bug patch PHP023 (generated by gen-patch, magma-agent)
# target:   php
# pinned:   e49be5f8e2903dbefa262847f1494c8e4c86004e
# upstream: 404e8bdb68 2022-09-09 "Fix #81726: phar wrapper: DOS when using quine gzip file"
# cve:      CVE-2022-31628
# class:    dos
# site:     ext/phar/phar.c:phar_open_from_fp
# canary:   recursion_count == 1
# harness:  execute (new)
# Lines above the first 'diff --git' are ignored by patch(1) and git-apply.
--- a/ext/phar/phar.c
+++ b/ext/phar/phar.c
@@ -1666,7 +1666,11 @@ static zend_result phar_open_from_fp(php_stream* fp, char *fname, size_t fname_l
 			MAPPHAR_ALLOC_FAIL("internal corruption of phar \"%s\" (truncated entry)")
 		}
 
+	#ifdef MAGMA_ENABLE_FIXES
 		if (!test && recursion_count) {
+	#else
+		if (!test) {
+	#endif
 			test = '\1';
 			pos = buffer+tokenlen;
 			if (!memcmp(pos, gz_magic, 3)) {
@@ -1730,10 +1734,17 @@ static zend_result phar_open_from_fp(php_stream* fp, char *fname, size_t fname_l
 
 				/* now, start over */
 				test = '\0';
+			#ifdef MAGMA_ENABLE_CANARIES
+				MAGMA_LOG("%MAGMA_BUG%", recursion_count == 1);
+			#endif
+			#ifdef MAGMA_ENABLE_FIXES
 				if (!--recursion_count) {
 					MAPPHAR_ALLOC_FAIL("unable to decompress gzipped phar archive \"%s\"");
 					break;
 				}
+			#else
+				--recursion_count;
+			#endif
 				continue;
 			} else if (!memcmp(pos, bz_magic, 3)) {
 				php_stream_filter *filter;
@@ -1773,10 +1784,14 @@ static zend_result phar_open_from_fp(php_stream* fp, char *fname, size_t fname_l
 
 				/* now, start over */
 				test = '\0';
+			#ifdef MAGMA_ENABLE_FIXES
 				if (!--recursion_count) {
 					MAPPHAR_ALLOC_FAIL("unable to decompress bzipped phar archive \"%s\"");
 					break;
 				}
+			#else
+				--recursion_count;
+			#endif
 				continue;
 			}
 
